### PR TITLE
Switch to batch uploading for cloud based mapping and add retries

### PR DESCRIPTION
**Pull request by cschuet**

### CODE REVIEW

File: cartographer/cloud/internal/handlers/add_sensor_data_batch_handler.cc [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        break;
+      case proto::SensorData::kLandmarkData:
+        GetUnsynchronizedContext<MapBuilderContextInterface>()
+            ->EnqueueSensorData(
+                sensor_data.sensor_metadata().trajectory_id(),
+                sensor::MakeDispatchable(
+                    sensor_data.sensor_metadata().sensor_id(),
+                    sensor::FromProto(sensor_data.landmark_data())));
+        break;
+      case proto::SensorData::kLocalSlamResultData: {
```
Comment (gaschler) [inline cartographer/cloud/internal/handlers/add_sensor_data_batch_handler.cc]: remove {}

File: cartographer/cloud/internal/local_trajectory_uploader.cc [hunk 2]
```diff
@@ -135,30 +93,19 @@ void LocalTrajectoryUploader::Shutdown() {
...
-        LOG(FATAL) << "Unknown message type: " << data_message->GetTypeName();
+    auto sensor_data = send_queue_.PopWithTimeout(kPopTimeout);
+    if (sensor_data) {
+      proto::SensorData *added_sensor_data = batch_request.add_sensor_data();
+      *added_sensor_data = *sensor_data;
+      TranslateTrajectoryId(added_sensor_data->mutable_sensor_metadata());
+      if (batch_request.sensor_data_size() == kBatchSize) {
+        async_grpc::Client<handlers::AddSensorDataBatchSignature> client(
+            client_channel_, async_grpc::CreateUnlimitedConstantDelayStrategy(
+                                 common::FromSeconds(1)));
```
Comment (gaschler) [inline cartographer/cloud/internal/local_trajectory_uploader.cc]: does this warn after a some retries?

File: cartographer/cloud/internal/local_trajectory_uploader.cc [hunk 3]
```diff
@@ -171,68 +118,6 @@ void LocalTrajectoryUploader::TranslateTrajectoryId(
...
-    add_local_slam_result_client_ = make_unique<
-        async_grpc::Client<handlers::AddLocalSlamResultDataSignature>>(
-        client_channel_);
-  }
-  TranslateTrajectoryId(data_request->mutable_sensor_metadata());
-  // A submap also holds a trajectory id that must be translated to uplink's
-  // trajectory id.
-  for (mapping::proto::Submap &mutable_submap :
-       *data_request->mutable_local_slam_result_data()->mutable_submaps()) {
-    mutable_submap.mutable_submap_id()->set_trajectory_id(
```
Comment (gaschler) [inline cartographer/cloud/internal/local_trajectory_uploader.cc]: Why is this no longer needed?

Comment (cschuet) [reply cartographer/cloud/internal/local_trajectory_uploader.cc]: Dude, you are awesome. Thanks for catching that!

File: cartographer/cloud/internal/client_server_test.cc [hunk 4]
```diff
@@ -281,6 +324,68 @@ TEST_F(ClientServerTest, GlobalSlam3D) {
...
+      local_slam_result_callback_);
+  TrajectoryBuilderInterface* trajectory_stub =
+      stub_for_uploading_server_->GetTrajectoryBuilder(trajectory_id);
+  const auto measurements = mapping::test::GenerateFakeRangeMeasurements(
+      kTravelDistance, kDuration, kTimeStep);
+  for (const auto& measurement : measurements) {
+    trajectory_stub->AddSensorData(kRangeSensorId.id, measurement);
+  }
+  WaitForLocalSlamResults(measurements.size());
+  WaitForLocalSlamResultUploads(6);
```
Comment (gaschler) [inline cartographer/cloud/internal/client_server_test.cc]: can you compute this?

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Pull request merged**